torchao/quantization/prototype/qat/affine_fake_quantized_tensor.py

```diff
@@ -0,0 +1,289 @@
+        preserve_zero: bool = True,
+        zero_point_domain: ZeroPointDomain = ZeroPointDomain.INT,
+    ):
+        def apply_fake_quant_fn(t: torch.Tensor):
+            assert isinstance(t, cls)
+            qmin, qmax = _get_and_check_qmin_qmax(target_dtype, quant_min, quant_max)
+            scale, zero_point = choose_qparams_affine(
+                t.original_tensor,
+                mapping_type,
+                block_size,
+                target_dtype,
+                qmin,
+                qmax,
+                eps,
+                scale_dtype,
+                zero_point_dtype,
+                preserve_zero,
+                zero_point_domain,
+            )
+            fq = _GenericFakeQuantize.apply(
+                t,
+                block_size,
+                scale,
+                zero_point,
+                qmin,
+                qmax,
+                zero_point_domain,
+            )
+            return fq
+        return cls(
+            input_float,
+            apply_fake_quant_fn,
+            fake_quant_enabled=True,
+            requires_grad=True,
+        )
+
+    def get_value(self) -> torch.Tensor:
+        if self.fake_quant_enabled:
+            return self.apply_fake_quant_fn(self)
+        else:
+            return _UnwrapAffineFakeQuantizedTensor.apply(self)
+
+    def _get_to_kwargs(self, *args, **kwargs):
+        device, dtype, _, memory_format = torch._C._nn._parse_to(*args, **kwargs)
+        device = self.device if device is None else device
+        dtype = self.dtype if dtype is None else dtype
+        memory_format = ( 
+            memory_format if memory_format is not None else torch.preserve_format
+        )   
+        kwargs = { 
+            "device": device,
+            "dtype": dtype,
+            "memory_format": memory_format,
+            "requires_grad": self.requires_grad,
+        }   
+        return kwargs
+
+    def to(self, *args, **kwargs):
+        kwargs = self._get_to_kwargs(*args, **kwargs)
+        device = kwargs.pop("device")
+        # not supported yet
+        kwargs.pop("memory_format")
+        return self.__class__(
+            self.original_tensor.to(device),
+            self.apply_fake_quant_fn,
+            self.fake_quant_enabled,
+            **kwargs,
+        )
+
+    def _apply_fn_to_data(self, fn: Callable):
+        """
+        Create a new `AffineFakeQuantizedTensor` with `fn` applied to the
+        original tensor, to be called within __torch_dispatch__.
+        """
+        return self._create_new(fn(self.original_tensor))
+
+    def _create_new(self, new_value: torch.Tensor):
+        """
+        Create a new `AffineFakeQuantizedTensor` with a new value,
+        to be called within __torch_dispatch__.
+        """
+        return self.__class__(
+            new_value,
+            self.apply_fake_quant_fn,
+            self.fake_quant_enabled,
+            requires_grad=False,
```

Comment:
so why is this set to `False` here? maybe adding a bit more detail in the docstring might help

Reply:
Yeah sure, will explain in more detail in the docstring. It's because this is called in `__torch_dispatch__`, which cannot produce tensors with gradients (because autograd will try to attach metadata after `__torch_dispatch__` and will error if it finds existing metadata).